File: sdks/php/src/Storage.php
<?php

declare(strict_types=1);

namespace Backlex;

/** File operations against /api/storage. */
final class Storage
{
    public function __construct(private Client $client)
    {
    }

    /** List stored objects, optionally filtered by key prefix. */
    public function list(?string $prefix = null): array
    {
        $path = '/api/storage';
        if ($prefix !== null && $prefix !== '') {
            $path .= '?prefix=' . rawurlencode($prefix);
        }
        return $this->client->request('GET', $path)['data'];
    }

    /** Upload bytes under $key. Pass contentType/folderId null to omit them. */
    public function put(string $key, string $body, ?string $contentType = null, ?string $folderId = null): array
    {
        $path = '/api/storage/' . rawurlencode($key);
        if ($folderId !== null && $folderId !== '') {
            $path .= '?folderId=' . rawurlencode($folderId);
        }
        [, $resp] = $this->client->sendRaw('PUT', $path, $body, $contentType);
        return ($resp === null || $resp === '') ? [] : json_decode($resp, true);
    }

    /** Fetch the raw bytes for $key. */
    public function download(string $key): string
    {
        [, $resp] = $this->client->sendRaw('GET', '/api/storage/' . rawurlencode($key), null, null);
        return $resp ?? '';
    }

    /** Remove the object at $key. */
    public function delete(string $key): array
    {
        return $this->client->request('DELETE', '/api/storage/' . rawurlencode($key));
    }
}

File: sdks/dart/lib/src/storage.dart
part of backlex;

/// File operations against `/api/storage`.
class Storage {
  final Client _client;

  Storage(this._client);

  /// List stored objects, optionally filtered by key prefix.
  Future<List<dynamic>> list([String? prefix]) async {
    var path = '/api/storage';
    if (prefix != null && prefix.isNotEmpty) {
      path += '?prefix=${Uri.encodeQueryComponent(prefix)}';
    }
    final r = await _client.request('GET', path) as Map<String, dynamic>;
    return r['data'] as List<dynamic>;
  }

  /// Upload bytes under [key]. Pass `contentType`/`folderId` null to omit them.
  Future<Map<String, dynamic>> put(String key, List<int> body,
      {String? contentType, String? folderId}) async {
    var path = '/api/storage/${Uri.encodeComponent(key)}';
    if (folderId != null && folderId.isNotEmpty) {
      path += '?folderId=${Uri.encodeQueryComponent(folderId)}';
    }
    final text = await _client.sendRaw('PUT', path, body, contentType);
    return (text.isEmpty ? <String, dynamic>{} : jsonDecode(text)) as Map<String, dynamic>;
  }

  /// Fetch the raw bytes for [key].
  Future<List<int>> download(String key) =>
      _client.downloadRaw('/api/storage/${Uri.encodeComponent(key)}');

  /// Remove the object at [key].
  Future<Map<String, dynamic>> delete(String key) async =>
      await _client.request('DELETE', '/api/storage/${Uri.encodeComponent(key)}') as Map<String, dynamic>;
}
